### PR TITLE
Release Zebra v1.0.1

### LINKED ISSUE #7021: Publish next Zebra release v1.0.1
# Prepare for the Release

 These release steps can be done a week before the release, in separate PRs.
 They can be skipped for urgent releases.

 ## Checkpoints

 For performance and security, we want to update the Zebra checkpoints in every release.
 - [x] You can copy the latest checkpoints from CI by following [the zebra-checkpoints README](https://github.com/ZcashFoundation/zebra/blob/main/zebra-utils/README.md#zebra-checkpoints).

 ## Missed Dependency Updates

<details>

<summary>This step will be skipped because we have an open ECC crate update PR.</summary>

 Sometimes `dependabot` misses some dependency updates, or we accidentally turned them off.

 This step can be skipped if there is a large pending dependency upgrade. (For example, shared ECC crates.)

 Here's how we make sure we got everything:
 - [ ] Run `cargo update` on the latest `main` branch, and keep the output
 - [ ] If needed, [add duplicate dependency exceptions to deny.toml](https://github.com/ZcashFoundation/zebra/blob/main/book/src/dev/continuous-integration.md#fixing-duplicate-dependencies-in-check-denytoml-bans)
 - [ ] If needed, remove resolved duplicate dependencies from `deny.toml`
 - [ ] Open a separate PR with the changes
 - [ ] Add the output of `cargo update` to that PR as a comment

</details>

 # Prepare and Publish the Release

 Follow the steps in the [release checklist](https://github.com/ZcashFoundation/zebra/blob/main/.github/PULL_REQUEST_TEMPLATE/release-checklist.md) to prepare the release:

 Release PR:
 - [x] Update Changelog
 - [x] Update README
 - [x] We need to decide whether this will be a 1.1.0 or 1.0.1 release
 - [x] Update Zebra Versions
 - [x] Update End of Support Height

 Publish Release:
 - [x] Create & Test GitHub Pre-Release
 - [x] Publish GitHub Release
 - [x] Publish Rust Crates
 - [x] Publish Docker Images

### PR DESCRIPTION
**Pull request by teor2345**
## Changelog Updates

- [x] Add these new user-visible PRs to the changelog:
  - #7045
  - #7095
  - #7102
  - #7103 (release blocker)
- [x] Check for other commits to main since `#7106` on 30 June: https://github.com/ZcashFoundation/zebra/commits/main

Optional:
  - [x] add entry for #7104 (in PR #7128)
  - [x] add entry for #7094

# Prepare for the Release

- [x] Make sure the PRs with the new checkpoint hashes ~~and missed dependencies~~ are already merged.
      (See the release ticket checklist for details)

# Summarise Release Changes

These steps can be done a few days before the release, in the same PR:

## Change Log

**Important**: Any merge into `main` deletes any edits to the draft changelog.
Once you are ready to tag a release, copy the draft changelog into `CHANGELOG.md`.

We use [the Release Drafter workflow](https://github.com/marketplace/actions/release-drafter) to automatically create a [draft changelog](https://github.com/ZcashFoundation/zebra/releases). We follow the [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) format.

To create the final change log:
- [x] Copy the **latest** draft changelog into `CHANGELOG.md` (there can be multiple draft releases)
- [x] Delete any trivial changes
    - [x] Put the list of deleted changelog entries in a PR comment to make reviewing easier
- [x] Combine duplicate changes
- [x] Edit change descriptions so they will make sense to Zebra users
- [x] Check the category for each change
  - Prefer the "Fix" category if you're not sure

## README

README updates can be skipped for urgent releases.

Update the README to:
- [x] Remove any "Known Issues" that have been fixed since the last release.
- [x] Update the "Build and Run Instructions" with any new dependencies.
      Check for changes in the `Dockerfile` since the last tag: `git diff <previous-release-tag> docker/Dockerfile`.
- [x] If Zebra has started using newer Rust language features or standard library APIs, update the known working Rust version in the README, book, and `Cargo.toml`s

You can use a command like:
```sh
fastmod --fixed-strings '1.58' '1.65'
```

## Create the Release PR

- [x] Push the updated changelog and README into a new branch
      for example: `bump-v1.0.0` - this needs to be different to the tag name
- [x] Create a release PR by adding `&template=release-checklist.md` to the comparing url ([Example](https://github.com/ZcashFoundation/zebra/compare/bump-v1.0.0?expand=1&template=release-checklist.md)).
- [x] Freeze the [`batched` queue](https://dashboard.mergify.com/github/ZcashFoundation/repo/zebra/queues) using Mergify.
- [x] Mark all the release PRs as `Critical` priority, so they go in the `urgent` Mergify queue.

# Update Versions and End of Support

### Choose a Release Level

Zebra follows [semantic versioning](https://semver.org). Semantic versions look like: MAJOR.MINOR.PATCH[-TAG.PRE-RELEASE]

Choose a release level for `zebrad`. Release levels are based on user-visible changes from the changelog:
- Mainnet Network Upgrades are `major` releases
- significant new features, large changes, deprecations, and removals are `minor` releases
- otherwise, it is a `patch` release

Zebra's Rust API doesn't have any support or stability guarantees, so we keep all the `zebra-*` and `tower-*` crates on a beta `pre-release` version.

### Update Crate Versions

<details>

<summary>If you're publishing crates for the first time:</summary>

- [ ] Install `cargo-release`: `cargo install cargo-release`
- [ ] Make sure you are  an owner of the crate or [a member of the Zebra crates.io `owners` group on GitHub](https://github.com/orgs/ZcashFoundation/teams/owners)

</details>

- [x] Update crate versions, commit the changes to the release branch, and do a release dry-run:

```sh
cargo clean # optional
cargo release version --verbose --execute --workspace --exclude zebrad beta
cargo release version --verbose --execute --package zebrad patch # [ major | minor | patch ]
cargo release commit --verbose --execute
# TODO: fix missing dependency errors in cargo release, so we can do a dry run for all the crates 
cargo release publish --verbose --dry-run --workspace --exclude zebra-consensus --exclude zebra-rpc --exclude zebra-utils --exclude zebrad
```

## Update End of Support

The end of support height is calculated from the current blockchain height:
- [x] Find where the Zcash blockchain tip is now by using a [Zcash explorer](https://zcashblockexplorer.com/blocks) or other tool.
- [x] Replace `ESTIMATED_RELEASE_HEIGHT` in [`end_of_support.rs`](https://github.com/ZcashFoundation/zebra/blob/main/zebrad/src/components/sync/end_of_support.rs) with the height you estimate the release will be tagged.

<details>

<summary>Optional: calculate the release tagging height</summary>

- Add `1152` blocks for each day until the release
- For example, if the release is in 3 days, add `1152 * 3` to the current Mainnet block height

</details>

## Update the Release PR

- [x] Push the version increments and the release constants to the release branch.

# Publish the Zebra Release

## Create the GitHub Pre-Release

- [x] Wait for all the release PRs to be merged
    - [x] #7103 (release blocker)
    - [x] #7090 (release preparation)
- [x] Create a new release using the draft release as a base, by clicking the Edit icon in the [draft release](https://github.com/ZcashFoundation/zebra/releases)
- [x] Set the tag name to the version tag,
      for example: `v1.0.0`
- [x] Set the release to target the `main` branch
- [x] Set the release title to `Zebra ` followed by the version tag,
      for example: `Zebra 1.0.0`
- [x] Replace the prepopulated draft changelog in the release description with the final changelog you created;
      starting just _after_ the title `## [Zebra ...` of the current version being released,
      and ending just _before_ the title of the previous release.
- [x] Mark the release as 'pre-release', until it has been built and tested
- [x] Publish the pre-release to GitHub using "Publish Release"
- [x] Delete all the [draft releases from the list of releases](https://github.com/ZcashFoundation/zebra/releases)

## Test the Pre-Release

- [x] Wait until the [Docker binaries have been built on `main`](https://github.com/ZcashFoundation/zebra/actions/workflows/continous-integration-docker.yml), and the quick tests have passed.
      (You can ignore the full sync and `lightwalletd` tests, because they take about a day to run.)
- [x] Wait until the [pre-release deployment machines have successfully launched](https://github.com/ZcashFoundation/zebra/actions/workflows/continous-delivery.yml)

## Publish Release

- [x] [Publish the release to GitHub](https://github.com/ZcashFoundation/zebra/releases) by disabling 'pre-release', then clicking "Set as the latest release"

## Publish Crates

- [x] Run `cargo login`
- [ ] Run `cargo clean` in the zebra repo (optional)
- [x] Publish the crates to crates.io: `cargo release publish --verbose --workspace --execute`
- [x] Check that Zebra can be installed from `crates.io`:
      `cargo install --locked --force --version 1.0.1 zebrad && ~/.cargo/bin/zebrad`
      and put the output in a comment on the PR.

## Publish Docker Images
- [x] Wait for the [the Docker images to be published successfully](https://github.com/ZcashFoundation/zebra/actions/workflows/release-binaries.yml).
- [x] Un-freeze the [`batched` queue](https://dashboard.mergify.com/github/ZcashFoundation/repo/zebra/queues) using Mergify.

## Release Failures

If building or running fails after tagging:

<details>

<summary>Tag a new release, following these instructions...</summary>

1. Fix the bug that caused the failure
2. Start a new `patch` release
3. Skip the **Release Preparation**, and start at the **Release Changes** step
4. Update `CHANGELOG.md` with details about the fix
5. Follow the release checklist for the new Zebra version

</details>

Close #7021

### CODE REVIEW

**Comment by teor2345** [comment]
Trivial changes:

```
$ cat trivial-issues.txt | sort -u
```

 - add(ci): Run release builds and production Docker image tests on pull requests ([#7055](https://github.com/ZcashFoundation/zebra/pull/7055))
 - add(tests): Add snapshot tests for sprout database formats ([#7057](https://github.com/ZcashFoundation/zebra/pull/7057))
 - build(deps): bump Swatinem/rust-cache from 2.4.0 to 2.5.0 ([#7002](https://github.com/ZcashFoundation/zebra/pull/7002))
 - build(deps): bump baptiste0928/cargo-install from 2.0.0 to 2.1.0 ([#6903](https://github.com/ZcashFoundation/zebra/pull/6903))
 - build(deps): bump bitflags from 2.3.1 to 2.3.2 ([#6943](https://github.com/ZcashFoundation/zebra/pull/6943))
 - build(deps): bump clap from 4.3.3 to 4.3.4 ([#6957](https://github.com/ZcashFoundation/zebra/pull/6957))
 - build(deps): bump clap from 4.3.4 to 4.3.5 ([#7038](https://github.com/ZcashFoundation/zebra/pull/7038))
 - build(deps): bump clap from 4.3.5 to 4.3.6 ([#7059](https://github.com/ZcashFoundation/zebra/pull/7059))
 - build(deps): bump clap from 4.3.6 to 4.3.8 ([#7066](https://github.com/ZcashFoundation/zebra/pull/7066))
 - build(deps): bump docker/build-push-action from 4.1.0 to 4.1.1 ([#6942](https://github.com/ZcashFoundation/zebra/pull/6942))
 - build(deps): bump docker/metadata-action from 4.5.0 to 4.6.0 ([#6941](https://github.com/ZcashFoundation/zebra/pull/6941))
 - build(deps): bump hyper from 0.14.26 to 0.14.27 ([#7082](https://github.com/ZcashFoundation/zebra/pull/7082))
 - build(deps): bump insta from 1.29.0 to 1.30.0 ([#7051](https://github.com/ZcashFoundation/zebra/pull/7051))
 - build(deps): bump itertools from 0.10.5 to 0.11.0 ([#7050](https://github.com/ZcashFoundation/zebra/pull/7050))
 - build(deps): bump log from 0.4.18 to 0.4.19 ([#6919](https://github.com/ZcashFoundation/zebra/pull/6919))
 - build(deps): bump peter-evans/dockerhub-description from 3.4.1 to 3.4.2 ([#7023](https://github.com/ZcashFoundation/zebra/pull/7023))
 - build(deps): bump reviewdog/action-actionlint from 1.37.0 to 1.37.1 ([#7003](https://github.com/ZcashFoundation/zebra/pull/7003))
 - build(deps): bump sentry from 0.31.3 to 0.31.4 ([#6958](https://github.com/ZcashFoundation/zebra/pull/6958))
 - build(deps): bump sentry from 0.31.4 to 0.31.5 ([#6988](https://github.com/ZcashFoundation/zebra/pull/6988))
 - build(deps): bump serde\_json from 1.0.96 to 1.0.97 ([#6987](https://github.com/ZcashFoundation/zebra/pull/6987))
 - build(deps): bump tj-actions/changed-files from 36.1.0 to 36.2.1 ([#6973](https://github.com/ZcashFoundation/zebra/pull/6973))
 - build(deps): bump tj-actions/changed-files from 36.2.1 to 36.3.0 ([#6986](https://github.com/ZcashFoundation/zebra/pull/6986))
 - build(deps): bump tj-actions/changed-files from 36.3.0 to 36.4.0 ([#7004](https://github.com/ZcashFoundation/zebra/pull/7004))
 - build(deps): bump tj-actions/changed-files from 36.4.0 to 36.4.1 ([#7024](https://github.com/ZcashFoundation/zebra/pull/7024))
 - build(deps): bump tj-actions/changed-files from 36.4.1 to 37.0.3 ([#7065](https://github.com/ZcashFoundation/zebra/pull/7065))
 - build(deps): bump w9jds/firebase-action from 11.30.1 to 12.4.0 ([#7005](https://github.com/ZcashFoundation/zebra/pull/7005))
 - change(commands): Ignore error from loading config if running the 'generate' or 'download' commands ([#7014](https://github.com/ZcashFoundation/zebra/pull/7014))
 - change(docs): Add `cargo clean` step to crate publishing steps ([#6959](https://github.com/ZcashFoundation/zebra/pull/6959))
 - change(docs): Explicitly invoke --execute when bumping crate versions ([#6949](https://github.com/ZcashFoundation/zebra/pull/6949))
 - change(network): Configurable maximum connections per IP ([#7013](https://github.com/ZcashFoundation/zebra/pull/7013))
 - change(release): Add deny.toml update details to release-checklist.md ([#7042](https://github.com/ZcashFoundation/zebra/pull/7042))
 - change(release): Add more cargo clean to the release checklist ([#6964](https://github.com/ZcashFoundation/zebra/pull/6964))
 - change(release): Change network upgrade wording in release-checklist.md ([#7010](https://github.com/ZcashFoundation/zebra/pull/7010))
 - change(rename): Update missed tower-batch-control renames ([#7011](https://github.com/ZcashFoundation/zebra/pull/7011))
 - change(state): Use `OrderedUtxo` in `CheckpointVerifiedBlock` ([#6971](https://github.com/ZcashFoundation/zebra/pull/6971))
 - cleanup(test): Make test debugging output more readable ([#7027](https://github.com/ZcashFoundation/zebra/pull/7027))
 - cleanup(workflow): Remove an outdated TODO in release-binaries.yml ([#6978](https://github.com/ZcashFoundation/zebra/pull/6978))
 - fix(build): Suppress warnings about doc links pointing to private items ([#6944](https://github.com/ZcashFoundation/zebra/pull/6944))
 - fix(cd): Rename a CD job with the same name as a CI job ([#7063](https://github.com/ZcashFoundation/zebra/pull/7063))
 - fix(changelog): Fix a broken link in CHANGELOG.md ([#6979](https://github.com/ZcashFoundation/zebra/pull/6979))
 - fix(cleanup): redundant configuration feature ([#6929](https://github.com/ZcashFoundation/zebra/pull/6929))
 - fix(deps): Replace openssl with rustls in tests and experimental features ([#7047](https://github.com/ZcashFoundation/zebra/pull/7047))
 - fix(doc): Add `fastmod --hidden` to mass-renames.md ([#6913](https://github.com/ZcashFoundation/zebra/pull/6913))
 - fix(docker): Improve Dockerfile cache use by making build commands match exactly ([#6933](https://github.com/ZcashFoundation/zebra/pull/6933))
 - fix(docker): Stop resetting the `cargo-chef` cache in the Dockerfile ([#6934](https://github.com/ZcashFoundation/zebra/pull/6934))
 - fix(logs): Avoid grouping logs for separate tasks ([#6923](https://github.com/ZcashFoundation/zebra/pull/6923))
 - fix(net): Clean up licensing, closure `move`, log typos, tracing spans ([#6995](https://github.com/ZcashFoundation/zebra/pull/6995))
 - fix(release): Use correct cargo release manifest key name ([#7028](https://github.com/ZcashFoundation/zebra/pull/7028))
 - fix(state): Replace a chain length assertion with a NotReadyToBeCommitted error ([#7072](https://github.com/ZcashFoundation/zebra/pull/7072))
 - refactor(app): De-duplicate and fix version handling code ([#6996](https://github.com/ZcashFoundation/zebra/pull/6996))
 - refactor(consensus): Rename router\_verifier to block\_verifier\_router ([#6998](https://github.com/ZcashFoundation/zebra/pull/6998))
 - rename(state): do additional renaming for clarification purposes ([#6967](https://github.com/ZcashFoundation/zebra/pull/6967))
- Refactor the structure of finalizable blocks, to make validation clearer ([#7035](https://github.com/ZcashFoundation/zebra/pull/7035), [#7025](https://github.com/ZcashFoundation/zebra/pull/7025))

**Comment by teor2345** [comment]
> * If Zebra has started using newer Rust language features or standard library APIs, update the known working Rust version in the README, book, and `Cargo.toml`s

This step was done in PR #7032.

**Comment by teor2345** [comment]
This is the draft changelog as of commit caf9a0925f80cd93003a90cad6394ded1bdc1b15 at Wednesday 28 June 0100 UTC.

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,80 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
...
+#### Network Security
+
+- Limit each IP address to 1 peer connection, to prevent denial of service attacks. ([#6980](https://github.com/ZcashFoundation/zebra/pull/6980), [#6993](https://github.com/ZcashFoundation/zebra/pull/6993))
+- Close new peer connections from the same IP and port, rather than replacing the older connection ([#6980](https://github.com/ZcashFoundation/zebra/pull/6980))
+- Reduce inbound service overloads and add a timeout ([#6950](https://github.com/ZcashFoundation/zebra/pull/6950))
+- Stop panicking when handling inbound connection handshakes ([#6984](https://github.com/ZcashFoundation/zebra/pull/6984))
+
+### Added
+
+- Make the maximum number of connections per IP configurable ([#7013](https://github.com/ZcashFoundation/zebra/pull/7013))
```
Comment (arya2) [inline CHANGELOG.md:48]: Nitpick: remove this from the omissions list
> change(network): Configurable maximum connections per IP (https://github.com/ZcashFoundation/zebra/pull/7013)

File: CHANGELOG.md [hunk 2]
```diff
@@ -6,6 +6,80 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
...
+
+### Support Changes
+
+These platforms are no longer supported by the Zebra team:
+- macOS has been moved from tier 2 to [tier 3 support](https://github.com/ZcashFoundation/zebra/blob/main/book/src/user/supported-platforms.md#tier-3) ([#6965](https://github.com/ZcashFoundation/zebra/pull/6965)). We disabled our regular macOS builds because Rust 1.70 [causes crashes during shutdown on macOS x86_64 (#6812)](https://github.com/ZcashFoundation/zebra/issues/6812). Zebra's state uses database transactions, so it should not be corrupted by the crash.
+
+### Security
+
+- Use Arc::into\_inner() to avoid potential hangs or panics ([#7032](https://github.com/ZcashFoundation/zebra/pull/7032))
+- Replace openssl with rustls in tests and experimental features ([#7047](https://github.com/ZcashFoundation/zebra/pull/7047))
```
Comment (arya2) [inline CHANGELOG.md:36]: Nitpick: remove this from the omissions list, add #7054, #7073, #7077, and #7088.

File: CHANGELOG.md [hunk 3]
```diff
@@ -6,6 +6,80 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
...
+#### Configuration
+
+- Ignore error from loading config if running the 'generate' or 'download' commands ([#7014](https://github.com/ZcashFoundation/zebra/pull/7014))
+- Apply force\_color to panic logs ([#6997](https://github.com/ZcashFoundation/zebra/pull/6997))
+
+#### Logging & Error Handling
+
+- Log a zebra-network task cancel on shutdown, rather than panicking ([#7078](https://github.com/ZcashFoundation/zebra/pull/7078))
+- Fix incorrect function spans in some logs ([#6923](https://github.com/ZcashFoundation/zebra/pull/6923), [#6995](https://github.com/ZcashFoundation/zebra/pull/6995))
+- Replace a state validation chain length assertion with a NotReadyToBeCommitted error ([#7072](https://github.com/ZcashFoundation/zebra/pull/7072))
```
Comment (arya2) [inline CHANGELOG.md:75]: Nitpick: remove these from the omissions list.

**Review by arya2 — COMMENTED** [review]
The changelog looks good to me, it's currently only missing #7045.

File: CHANGELOG.md (hunk 1 above)
Comment (teor2345) [reply CHANGELOG.md:48]: Thanks, but I'm not going to keep that comment up to date.

File: CHANGELOG.md (hunk 2 above)
Comment (teor2345) [reply CHANGELOG.md:36]: Thanks, but I'm not going to keep that comment up to date, people can check recent commits if they want.

**Review by arya2 — APPROVED** [review]
The changelog and version bumps look good, and I agree with updating zebrad's patch version.

CI panicked in `wait_for_zebra_checkpoints_generation()`: https://github.com/ZcashFoundation/zebra/actions/runs/5418491535/jobs/9851372930?pr=7090#step:8:219

**Comment by teor2345** [comment]
```
$ cargo install --locked --force --version 1.0.1 zebrad
Downloaded zebrad v1.0.1
Downloaded 1 crate (231.9 KB) in 1.72s
Updating crates.io index
Installing zebrad v1.0.1
Updating crates.io index
warning: package `x25519-dalek v2.0.0-rc.2` in Cargo.lock is yanked in registry `crates-io`, consider running without --locked
Updating crates.io index
Compiling proc-macro2 v1.0.60
Compiling quote v1.0.28
...
   Compiling zebrad v1.0.1
   Compiling zcash_primitives v0.11.0
   Compiling zebra-chain v1.0.0-beta.27
   Compiling zcash_proofs v0.11.0
   Compiling zebra-node-services v1.0.0-beta.27
   Compiling zebra-network v1.0.0-beta.27
   Compiling zebra-script v1.0.0-beta.27
   Compiling rocksdb v0.21.0
   Compiling zebra-state v1.0.0-beta.27
   Compiling zebra-consensus v1.0.0-beta.27
   Compiling zebra-rpc v1.0.0-beta.27
    Finished release [optimized + debuginfo] target(s) in 3m 00s
   Replacing /home/dev/.cargo/bin/zebrad
    Replaced package `zebrad v1.0.0` with `zebrad v1.0.1` (executable `zebrad`)
```

```
$ ~/.cargo/bin/zebrad
```
Screenshot 2023-07-03 at 13 06 06

**Pull request merged**